### PR TITLE
Remove QMSI sensor subsystem ADC driver and use Designware driver for ARC.

### PR DESCRIPTION
**Pull request by punitvara**
Remove adc_qmsi_ss and consolidate code with adc_dw.c

### CODE REVIEW

File: drivers/adc/adc_dw.c [hunk 1]
```diff
@@ -206,7 +206,7 @@ static void adc_dw_disable(struct device *dev)
 	struct adc_info *info = dev->driver_data;
 	const struct adc_config *config = dev->config->config_info;
 	u32_t adc_base = config->reg_base;
-
+	printk("HELLo");
```
Comment (nashif) [inline drivers/adc/adc_dw.c]: Can you please remove this patch?

**Comment by nashif** [comment]
resolve the conflicts please

**Comment by nashif** [comment]
we still have a conflict, please ALWAYS rebase on top of master when submitting new changes.

**Comment by MaureenHelm** [comment]
> we still have a conflict, please ALWAYS rebase on top of master when submitting new changes.

I fixed the conflict in #7691

**Comment by punitvara** [comment]
@MaureenHelm Thanks I will update my PR

File: drivers/adc/adc_dw.c [hunk 2]
```diff
@@ -51,6 +54,19 @@
 #endif
 static void adc_config_irq(void);
 
+
+#ifdef CONFIG_ADC_DW
```
Comment (nashif) [inline drivers/adc/adc_dw.c]: why are you checking for CONFIG_ADC_DW here? This driver is built when CONFIG_ADC_DW is defined, so this becomes useless.

**Review by nashif — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_dw.c [hunk 3]
```diff
@@ -176,8 +193,28 @@ static void adc_goto_deep_power_down(void)
 		irq_unlock(state);
 		do {
 			reg_value = sys_in32(PERIPH_ADDR_BASE_CREG_SLV0);
+			/*Don't hang system if power down fails*/
+			if (i++ >= 10) {
+				printk("DW adc power down failed\n");
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: sure.

File: drivers/grove/light_sensor.c [hunk 4]
```diff
@@ -71,15 +81,22 @@ static int gls_init(struct device *dev)
...
-	drv_data->sample.sampling_delay = 12;
-	drv_data->sample.channel_id = CONFIG_GROVE_LIGHT_SENSOR_ADC_CHANNEL;
-	drv_data->sample.buffer = drv_data->adc_buffer;
-	drv_data->sample.buffer_length = 4;
-
-	drv_data->adc_table.entries = &drv_data->sample;
-	drv_data->adc_table.num_entries = 1;
-
-	adc_enable(drv_data->adc);
+#if defined(CONFIG_ARC)
```
Comment (nashif) [inline drivers/grove/light_sensor.c]: you should only ifdef the items that are different, not the whole thing.

**Review by nashif — CHANGES_REQUESTED** [review]

File: arch/arc/soc/quark_se_c1000_ss/dts.fixup [hunk 5]
```diff
@@ -68,4 +68,10 @@
 #define CONFIG_GPIO_QMSI_1_IRQ		INTEL_QMSI_GPIO_B0800B00_IRQ_0
 #define CONFIG_GPIO_QMSI_1_IRQ_PRI	INTEL_QMSI_GPIO_B0800B00_IRQ_0_PRIORITY
 
+#define CONFIG_ADC_0_IRQ		INTEL_QMSI_ADC_80015000_IRQ_NORMAL
```
Comment (nashif) [inline arch/arc/soc/quark_se_c1000_ss/dts.fixup]: we are moving away from QMSI, so why are you using INTEL_QMSI here? this is generic designware IP!

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 6]
```diff
@@ -176,5 +176,18 @@
 
 			status = "disabled";
 		};
+
+		adc0: adc@80015000 {
+			compatible = "intel,qmsi-adc";
```
Comment (nashif) [inline dts/arc/quark_se_c1000_ss.dtsi]: this is neither intel, not qmsi, this is designware IP

File: dts/bindings/iio/adc/intel,qmsi-adc.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+inherits:
+    !include adc.yaml
+
+properties:
+    compatible:
+      type: string
+      category: required
+      description: compatible strings
+      constraint: "intel,qmsi-adc"
```
Comment (nashif) [inline dts/bindings/iio/adc/intel,qmsi-adc.yaml]: not intel and not qmsi, this is designware

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by ManojSubbarao** [comment]
@punitvara  Please address new review comments to drive this to closure at earliest.

File: arch/arc/soc/quark_se_c1000_ss/dts.fixup (hunk 5 above)
Comment (punitvara) [reply arch/arc/soc/quark_se_c1000_ss/dts.fixup]: I will change the name to DW_ADC_80015000_IRQ_NORMAL. Hope its fine

File: dts/bindings/iio/adc/intel,qmsi-adc.yaml (hunk 7 above)
Comment (punitvara) [reply dts/bindings/iio/adc/intel,qmsi-adc.yaml]: ok

File: drivers/grove/light_sensor.c (hunk 4 above)
Comment (punitvara) [reply drivers/grove/light_sensor.c]: For ARC board whatever was present, that's enough for now. If any extra parameters such as GAIN is required for other boards, we can add it later with if def

File: drivers/adc/adc_dw.c [hunk 8]
```diff
@@ -176,8 +191,28 @@ static void adc_goto_deep_power_down(void)
...
 		} while ((reg_value & 0x1) == 0);
 	}
+
+}
+
+static void set_resolution(struct device *dev,
+			  const struct adc_sequence *sequence)
+{
+	u32_t tmp_val;
+	u32_t adc_base = CONFIG_ADC_0_BASE_ADDRESS;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: ok.

File: drivers/adc/adc_dw.c [hunk 9]
```diff
@@ -213,75 +248,69 @@ static void adc_dw_disable(struct device *dev)
...
-	ctrl |= ADC_SEQ_PTR_RST;
-	sys_out32(ctrl, adc_base + ADC_CTRL);
-
-	tmp_val = sys_in32(adc_base + ADC_SET);
-	tmp_val &= ADC_SEQ_SIZE_SET_MASK;
-	tmp_val |= (((seq_tbl->num_entries - 1) & SIX_BITS_SET)
-		<< SEQ_ENTRIES_POS);
-	tmp_val |= ((seq_tbl->num_entries - 1) << THRESHOLD_POS);
-	sys_out32(tmp_val, adc_base + ADC_SET);
+	adc_info_dev.active_channels = 1 << channel_id;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: ok

File: drivers/adc/adc_dw.c [hunk 10]
```diff
@@ -338,99 +413,43 @@ int adc_dw_init(struct device *dev)
...
 }
 
-#ifdef CONFIG_ADC_DW_SINGLESHOT
 static void adc_dw_rx_isr(void *arg)
 {
 	struct device *dev = (struct device *)arg;
 	struct device_config *dev_config = dev->config;
 	const struct adc_config *config = dev_config->config_info;
 	struct adc_info *info = dev->driver_data;
 	u32_t adc_base = config->reg_base;
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: ok

File: drivers/adc/adc_dw.c [hunk 11]
```diff
@@ -291,24 +320,71 @@ static int adc_dw_read_request(struct device *dev, struct adc_seq_table *seq_tbl
...
+	adc_dw_enable(adc_info_dev.dev);
+	ctrl = sys_in32(adc_base + ADC_CTRL);
+	ctrl |= ADC_SEQ_PTR_RST;
+	sys_out32(ctrl, adc_base + ADC_CTRL);
+
+	tmp_val = sys_in32(adc_base + ADC_SET);
+	tmp_val &= ADC_SEQ_SIZE_SET_MASK;
+	tmp_val |= ((0 & SIX_BITS_SET)
+		<< SEQ_ENTRIES_POS);
+	tmp_val |= (0 << THRESHOLD_POS);
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: ok. agreed

File: drivers/adc/adc_dw.c [hunk 12]
```diff
@@ -213,75 +248,69 @@ static void adc_dw_disable(struct device *dev)
...
 
-	if ((seq_tbl->num_entries % 2) != 0) {
-		tmp_val = ((entry[0].sampling_delay & ELEVEN_BITS_SET)
-			<< SEQ_DELAY_EVEN_POS);
-		tmp_val |= (entry[0].channel_id & FIVE_BITS_SET);
-		sys_out32(tmp_val, adc_base + ADC_SEQ);
+	/* multiple channels not supported to read, if bitmask "channels"
+	 * has more than 1 bit set than return error
+	 */
+	if ((hit > 1) || !(seq_tbl->channels & adc_info_dev.active_channels)) {
```
[the remark this answers is not in the document]
Comment (punitvara) [reply drivers/adc/adc_dw.c]: agreed

**Pull request closed without merging**